himbaechel/uarch/gowin/pack.cc

```diff
@@ -1941,6 +1941,98 @@ struct GowinPacker
         return dummy_ci;
     }
 
+    // optimize ALU wiring
+    // A very simple ALU optimization: once we detect that one of the inputs is
+    // a constant, we modify the main LUT that describes the ALU function so
+    // that this primitive input is ignored, and then disconnect it from the
+    // network, freeing up the PIP.
+    // For example (unrealistic, since a real ALU LUT has a larger size and
+    // service bits in the middle, etc.), the addition function of A and B when
+    // A = 1 is converted from the general case (A isn't a constant and B isn't a
+    // constant) to a special case:
+    // 0110 -> 0011
+    void optimize_alu_lut(CellInfo *ci, int mode)
+    {
+        auto uni_shift = [&](unsigned int val, int amount) {
+            if (amount < 0) {
+                return val >> -amount;
+            }
+            return val << amount;
+        };
+
+        IdString vcc_net_name = ctx->id("$PACKER_VCC");
+        IdString gnd_net_name = ctx->id("$PACKER_GND");
+        bool optimized = false;
+        switch (mode) {
+        case 2: {
+            // ALU LUT for mode 2 is 0110_0000_1001_1010 for all chips
+            // We will change this feature if the next
+            // unreleased Gowin chip series changes this
+            // representation.
+            // If ADDSUB dynamically switches between + and -,
+            // optimization is not possible.
+            int possible_carry = 0b1100U;
+            IdString inp_net_name = ci->getPort(id_I3)->name;
+            if (inp_net_name != vcc_net_name && inp_net_name != gnd_net_name) {
+                break;
+            }
+            if (inp_net_name == gnd_net_name) {
+                possible_carry = 0b0011U;
+            }
+            unsigned int alu_lut = 0b0110000010011010U;
+            for (int i = 0; i < 3; ++i) {
+                if (i == 2) {
+                    break;
+                }
+                IdString inp_name = ctx->idf("I%d", i);
+                inp_net_name = ci->getPort(inp_name)->name;
+                if (inp_net_name == vcc_net_name || inp_net_name == gnd_net_name) {
+                    ci->disconnectPort(inp_name);
+                    optimized = true;
+
+                    // fix the carry
+                    if (!i) {
```

Comment:
nitpick - can this be an `if (i == 0)` because that makes the logic here a bit clearer imo